File: test/mem64/memtest.c
// clang --target=wasm64-freestanding -mbulk-memory -nostdlib -O2 -Wl,--no-entry -Wl,--export-dynamic -o memtest.wasm memtest.c

#include <stddef.h>
#include <stdint.h>

#define KB ((size_t)1024)
#define MB (1024 * KB)
#define GB (1024 * MB)

#define PAGE_SIZE (64 * KB)
#define PAGES_PER_GB (GB / PAGE_SIZE)

void assert(int value)
{
	if (value == 0)
	{
		__builtin_trap();
	}
}

__attribute__((visibility("default"))) void memtest(int32_t val_i32, int64_t val_i64, float val_f32, double val_f64)
{
	__builtin_wasm_memory_grow(0, PAGES_PER_GB * 8); // memory.grow
	char* mem = (char*)(GB * 4);
	char* mem_stores = (char*)(GB * 5);
	char* mem_loads = (char*)(GB * 6);

	size_t num_pages = __builtin_wasm_memory_size(0); // memory.size
	assert(num_pages == (PAGES_PER_GB * 8));

	*(int32_t*)(mem_stores + 0) = *(int32_t*)(mem_loads + 0); // i32.load -> i32.store
	*(int64_t*)(mem_stores + 8) = *(int64_t*)(mem_loads + 8); // i64.load -> i64.store
	*(float*)(mem_stores + 16)  = *(float*)(mem_loads + 16); // f32.load -> f32.store
	*(double*)(mem_stores + 24) = *(double*)(mem_loads + 24); // f64.load -> f64.store

	*(int8_t*)(mem_stores + 32) = (int8_t)val_i32; // i32.store8
	*(int16_t*)(mem_stores + 40) = (int16_t)val_i32; // i32.store16
	*(int8_t*)(mem_stores + 48) = (int8_t)val_i64; // i64.store8
	*(int16_t*)(mem_stores + 56) = (int16_t)val_i64; // i64.store16
	*(int32_t*)(mem_stores + 64) = (int32_t)val_i64; // i64.store32

	int32_t store32 = 0;
	store32 += (int32_t)*(int8_t*)(mem_loads + 32); // i32.load8_s
	store32 += (int32_t)*(uint8_t*)(mem_loads + 40); // i32.load8_u
	store32 += (int32_t)*(int16_t*)(mem_loads + 48); // i32.load16_s
	store32 += (int32_t)*(uint16_t*)(mem_loads + 56); // i32.load16_s

	int64_t store64 = 0;
	store64 += (int64_t)*(int8_t*)(mem_loads + 64); // i64.load8_s
	store64 += (int64_t)*(uint8_t*)(mem_loads + 72); // i64.load8_u
	store64 += (int64_t)*(int16_t*)(mem_loads + 80); // i64.load16_s
	store64 += (int64_t)*(uint16_t*)(mem_loads + 88); // i64.load16_s
	store64 += (int64_t)*(int32_t*)(mem_loads + 96); // i64.load32_s
	store64 += (int64_t)*(uint32_t*)(mem_loads + 104); // i64.load32_s

	// forces the compiler to not elide or condense the loads
	*(int32_t*)(mem_stores + 0) = store32;
	*(int64_t*)(mem_stores + 8) = store64;
	
	__builtin_memset(mem + KB, 0xFF, KB); // memory.fill
	__builtin_memcpy(mem + KB * 4, mem + KB * 3, KB); // memory.copy
}

File: test/mem64/memtest.zig
const KB = 1024;
const MB = 1024 * KB;
const GB = 1024 * MB;

const PAGE_SIZE = 64 * KB;
const PAGES_PER_GB = GB / PAGE_SIZE;

fn assert(cond: bool) !void {
    if (!cond) {
        return error.Failed;
    }
}

export fn memtest(val_i32: i32, val_i64: i64, val_f32: f32, val_f64: f64) i32 {
    testInternal(val_i32, val_i64, val_f32, val_f64) catch {
        return 1;
    };
    return 0;
}

fn testInternal(val_i32: i32, val_i64: i64, val_f32: f32, val_f64: f64) !void {
    _ = @wasmMemoryGrow(0, PAGES_PER_GB * 4);

    const grow_value: isize = @wasmMemoryGrow(0, PAGES_PER_GB * 6); // memory.grow
    try assert(grow_value != -1);
    const start_page: [*]volatile u8 = @ptrFromInt(@as(usize, @intCast(grow_value)));

    const mem = start_page + (GB * 4);
    const mem_stores = mem + MB * 1; // volatile?
    const mem_loads = mem + MB * 2; // volatile?

    const num_pages: usize = @wasmMemorySize(0);
    try assert(num_pages >= PAGES_PER_GB * 6);

    const ptr_load_i32 = @as(*volatile i32, @ptrCast(@alignCast(mem_loads)));
    const ptr_load_i64 = @as(*volatile i64, @ptrCast(@alignCast(mem_loads + 8)));
    const ptr_load_f32 = @as(*volatile f32, @ptrCast(@alignCast(mem_loads + 16)));
    const ptr_load_f64 = @as(*volatile f64, @ptrCast(@alignCast(mem_loads + 24)));

    ptr_load_i32.* = val_i32; // i32.store
    ptr_load_i64.* = val_i64; // i64.store
    ptr_load_f32.* = val_f32; // f32.store
    ptr_load_f64.* = val_f64; // f64.store

    try assert(ptr_load_i32.* == val_i32);
    try assert(ptr_load_i64.* == val_i64);
    try assert(ptr_load_f32.* == val_f32);
    try assert(ptr_load_f64.* == val_f64);

    const ptr_store_i32 = @as(*volatile i32, @ptrCast(@alignCast(mem_stores)));
    const ptr_store_i64 = @as(*volatile i64, @ptrCast(@alignCast(mem_stores + 8)));
    const ptr_store_f32 = @as(*volatile f32, @ptrCast(@alignCast(mem_stores + 16)));
    const ptr_store_f64 = @as(*volatile f64, @ptrCast(@alignCast(mem_stores + 24)));

    ptr_store_i32.* = ptr_load_i32.*; // i32.load && i32.store
    ptr_store_i64.* = ptr_load_i64.*; // i64.load && i64.store
    ptr_store_f32.* = ptr_load_f32.*; // f32.load && f32.store
    ptr_store_f64.* = ptr_load_f64.*; // f64.load && f64.store

    try assert(ptr_store_i32.* == ptr_load_i32.*);
    try assert(ptr_store_i64.* == ptr_load_i64.*);
    try assert(ptr_store_f32.* == ptr_load_f32.*);
    try assert(ptr_store_f64.* == ptr_load_f64.*);

    var load32: i32 = 0;
    ptr_load_i32.* = 0x7F;
    load32 = @as(*volatile i8, @ptrCast(@alignCast(ptr_load_i32))).*; // i32.load8_s
    try assert(load32 == 0x7F);
    ptr_load_i32.* = 0xFF;
    load32 = @as(*volatile u8, @ptrCast(@alignCast(ptr_load_i32))).*; // i32.load8_u
    try assert(load32 == 0xFF);
    ptr_load_i32.* = 0x7FFF;
    load32 = @as(*volatile i16, @ptrCast(@alignCast(ptr_load_i32))).*; // i32.load16_s
    try assert(load32 == 0x7FFF);
    ptr_load_i32.* = 0xFFFF;
    load32 = @as(*volatile u16, @ptrCast(@alignCast(ptr_load_i32))).*; // i32.load16_s
    try assert(load32 == 0xFFFF);

    var load64: i64 = 0;
    ptr_load_i64.* = 0x7F;
    load64 = @as(*volatile i8, @ptrCast(@alignCast(ptr_load_i64))).*; // i64.load8_s
    try assert(load64 == 0x7F);
    ptr_load_i64.* = 0xFF;
    load64 = @as(*volatile u8, @ptrCast(@alignCast(ptr_load_i64))).*; // i64.load8_u
    try assert(load64 == 0xFF);
    ptr_load_i64.* = 0x7FFF;
    load64 = @as(*volatile i16, @ptrCast(@alignCast(ptr_load_i64))).*; // i64.load16_s
    try assert(load64 == 0x7FFF);
    ptr_load_i64.* = 0xFFFF;
    load64 = @as(*volatile u16, @ptrCast(@alignCast(ptr_load_i64))).*; // i64.load16_s
    try assert(load64 == 0xFFFF);
    ptr_load_i64.* = 0x7FFFFFFF;
    load64 = @as(*volatile i32, @ptrCast(@alignCast(ptr_load_i64))).*; // i64.load32_s
    try assert(load64 == 0x7FFFFFFF);
    ptr_load_i64.* = 0xFFFFFFFF;
    load64 = @as(*volatile u32, @ptrCast(@alignCast(ptr_load_i64))).*; // i64.load32_s
    try assert(load64 == 0xFFFFFFFF);

    const memset_dest = (mem + KB)[0..KB];
    const memcpy_dest = (mem + KB * 2)[0..KB];
    @memset(memset_dest, 0xFF); // memory.fill
    @memcpy(memcpy_dest, memset_dest); // memory.copy

    try assert(memset_dest[0] == 0xFF);
    try assert(memset_dest[KB - 1] == 0xFF);
    try assert(memcpy_dest[0] == 0xFF);
    try assert(memcpy_dest[KB - 1] == 0xFF);
}
